### ISSUE TITLE
Phantomjs Memory Leak Issue

### ISSUE THREAD

**Issue by entrptaher** [ariya/phantomjs#14308]
1. Which version of PhantomJS are you using? Tip: run `phantomjs --version`.
   2.1.1 (but it doesn't matter, it happens to any version of Phantomjs)
2. What steps will reproduce the problem?
   1. Use any testing suite, ie. CasperJS to do a Ajax call or any kind of Async steps.
   2. Give it command while it's still running the previous step.
   3. Bam, memory leak.
   
   ![](http://i.imgur.com/n9YmIeU.png)
   ![Example Leak](http://i.imgur.com/ojl3FWx.png)
3. Which operating system are you using?
   I tried:
   - Ubuntu 12.04, 14.04, 16.04
   - Windows Xp, Windows 8.1
   - Digitalocean droplets (Ubuntu server 14.04 x64 image).
   
   All OS were both 32bit and 64 bit.
4. Did you use binary PhantomJS or did you compile it from source?
   I used Phantomjs from NPM, also built it, also tried older versions etc etc.
5. Please provide any additional information below.
   This is a general issue that has been talked over and over for past few years and yet no solution was provided.
   Where: 
   - https://github.com/ariya/phantomjs/issues/12903
   - https://github.com/ariya/phantomjs/issues/11390
   - https://github.com/karma-runner/karma-phantomjs-launcher/issues/55
   - https://github.com/ariya/phantomjs/issues/13027
   - https://github.com/ariya/phantomjs/issues/12903
   - https://github.com/ariya/phantomjs/issues/12317
   - https://github.com/angular/material/issues/4734
   - and many more
6. This has been talked over and over, why did you open another issue?
   Well, did any of them solve the issue? Lets solve this issue once and for all.
7. What can we do?
   For example, just when the leak happens, find out what it was doing, in my example, Let it wait for reply after a Ajax call. The Ajax call can be async, but waiting for the reply should be sync, thus preventing another command to execute while the ajax call is still in process.
   
   In CasperJS, there is a method called **waitFor** which can do the following, and the memory doesn't leak, _yet_. It might leak in someone elses computer.
   
   ``` js
   casper.waitFor(function check() {
   data = this.evaluate(function() {
       return document.querySelectorAll('html').length > 0;
   });
   return data;
   }, function then() {
   this.capture('yoursitelist.png');
   });
   ```

Now, I wonder, what can be done to prevent the memory leak.

**Comment by byaka** [comment]
@entrptaher, great work!
Hope this bug will be fixed in future...

**Comment by entrptaher** [comment]
In the webserver module, if the concurrent number of connections can be adjusted/limited to 1, then maybe all of the steps would be blocked by default.

**Comment by jwoertink** [comment]
I think I'm seeing a memory leak too. I'm using phantomjs version 2.1.1 on ubuntu 14.04. I open the browser, make the request, then close the browser, then find the PID that it was using and kill that, but the process never seems to go away. Within 30 minutes, I have over 100 phantomjs processes running. 

This doesn't happen when running on OSX that I'm seeing.

**Comment by byaka** [comment]
@jwoertink this not look like mem-leak))
try to kill it with ```kill -9 [pid]```

**Comment by liuyihan** [comment]
Also I'm seeing this issue, hope to be fixed in future…I'm using phantomjs version 2.1.1 on ubuntu 14.04. After I make 10000 request,  the process hold almost 1GB memory which cannot be released.

**Comment by entrptaher** [comment]
So the development stopped at 2.1.1.
That's sad I guess.

**Comment by Xsmael** [comment]
I'm facing the same, problem! any hope that this issue will get solved ?

Thanks.

**Comment by Vitallium** [comment]
Let's track this in https://github.com/ariya/phantomjs/issues/11390
